IPFIX/NetFlow-v9 flows without flow-start/end fields fail to insert (null in non-nullable firstSwitched)

## Summary

When an IPFIX or NetFlow v9 exporter sends flows without any flow-start/end timing elements, insertion into ClickHouse fails:

```
WARN ParserBase : Error preparing records for dispatch.
Caused by: java.lang.IllegalArgumentException: An attempt to write null into not nullable column 'firstSwitched'
```

The flow is dropped; no data is stored for those exporters.

## Root cause

`firstSwitched`, `lastSwitched`, and `deltaSwitched` are non-nullable `DateTime64(9)` columns, but the builders can produce null:

- `IpFixFlowBuilder.getFirstSwitched()`/`getLastSwitched()` return **null** when no `flowStart*`/`flowEnd*` element is present.
- `Netflow9FlowBuilder` returns null when neither `FIRST_SWITCHED`+bootTime nor `flowStart/EndMilliseconds` is present.
- `getDeltaSwitched()` derives from those, so it can be null too.

`Flow.getFirstSwitched()` returns a non-`Optional` `Instant` (non-null contract), and sFlow/NetFlow-v5 already default to receipt/boot-derived times — IPFIX and NetFlow v9 don't, violating that contract.

## Why tests missed it

Every existing builder test sets the timing fields, so the no-timing path was never exercised.

## Fix

IPFIX and NetFlow v9 builders fall back to `receivedAt` when no flow-start/end is exported (matching sFlow), and `getDeltaSwitched()` falls back to `getFirstSwitched()` when the timeout calc yields null — so all three columns are always non-null. Added builder tests for the no-timing case.

## Workaround (existing installs)

Make the columns nullable:
```sql
ALTER TABLE riptide.flows
  MODIFY COLUMN firstSwitched Nullable(DateTime64(9)),
  MODIFY COLUMN lastSwitched  Nullable(DateTime64(9)),
  MODIFY COLUMN deltaSwitched Nullable(DateTime64(9));
```

--- a/src/main/java/org/riptide/flows/parser/ipfix/IpFixFlowBuilder.java
+++ b/src/main/java/org/riptide/flows/parser/ipfix/IpFixFlowBuilder.java
@@ -125,9 +125,10 @@ public class IpFixFlowBuilder {
                             if (rawFlow.flowStartSysUpTime != null) {
                                 return rawFlow.systemInitTimeMilliseconds.plus(rawFlow.flowStartSysUpTime);
                             }
-                            // No flow-start element exported: fall back to receipt time (as sFlow
-                            // does), honouring the non-null Flow contract and the non-nullable column.
-                            return receivedAt;
+                            // No flow-start element exported: fall back to the export time (the
+                            // packet header timestamp), as goflow2 does — honouring the non-null Flow
+                            // contract and the non-nullable column.
+                            return this.getTimestamp();
                         });
             }
 
@@ -145,8 +146,8 @@ public class IpFixFlowBuilder {
                             if (rawFlow.flowEndSysUpTime != null) {
                                 return rawFlow.systemInitTimeMilliseconds.plus(rawFlow.flowEndSysUpTime);
                             }
-                            // No flow-end element exported: fall back to receipt time (see getFirstSwitched).
-                            return receivedAt;
+                            // No flow-end element exported: fall back to the export time (see getFirstSwitched).
+                            return this.getTimestamp();
                         });
             }
 

--- a/src/main/java/org/riptide/flows/parser/ipfix/IpFixFlowBuilder.java
+++ b/src/main/java/org/riptide/flows/parser/ipfix/IpFixFlowBuilder.java
@@ -125,7 +125,9 @@ public class IpFixFlowBuilder {
                             if (rawFlow.flowStartSysUpTime != null) {
                                 return rawFlow.systemInitTimeMilliseconds.plus(rawFlow.flowStartSysUpTime);
                             }
-                            return null;
+                            // No flow-start element exported: fall back to receipt time (as sFlow
+                            // does), honouring the non-null Flow contract and the non-nullable column.
+                            return receivedAt;
                         });
             }
 
@@ -143,7 +145,8 @@ public class IpFixFlowBuilder {
                             if (rawFlow.flowEndSysUpTime != null) {
                                 return rawFlow.systemInitTimeMilliseconds.plus(rawFlow.flowEndSysUpTime);
                             }
-                            return null;
+                            // No flow-end element exported: fall back to receipt time (see getFirstSwitched).
+                            return receivedAt;
                         });
             }
 
@@ -152,7 +155,7 @@ public class IpFixFlowBuilder {
                 final var flowActiveTimeout = Optionals.first(rawFlow.flowActiveTimeout, flowActiveTimeoutFallback).orElse(null);
                 final var flowInactiveTimeout = Optionals.first(rawFlow.flowInactiveTimeout, flowInactiveTimeoutFallback).orElse(null);
 
-                return new Timeout()
+                final var delta = new Timeout()
                         .withActiveTimeout(flowActiveTimeout)
                         .withInactiveTimeout(flowInactiveTimeout)
                         .withFirstSwitched(this.getFirstSwitched())
@@ -160,6 +163,9 @@ public class IpFixFlowBuilder {
                         .withNumBytes(this.getBytes())
                         .withNumPackets(this.getPackets())
                         .calculateDeltaSwitched();
+                // The timeout calc can yield null (no timeouts); default to firstSwitched like the
+                // Flow interface, which is now non-null.
+                return delta != null ? delta : this.getFirstSwitched();
             }
 
             @Override

--- a/src/main/java/org/riptide/flows/parser/netflow9/Netflow9FlowBuilder.java
+++ b/src/main/java/org/riptide/flows/parser/netflow9/Netflow9FlowBuilder.java
@@ -81,9 +81,12 @@ public class Netflow9FlowBuilder {
 
             @Override
             public Instant getFirstSwitched() {
-                return Optionals.of(raw.FIRST_SWITCHED)
+                final var firstSwitched = Optionals.of(raw.FIRST_SWITCHED)
                         .map(this.getBootTime()::plus)
                         .orElse(raw.flowStartMilliseconds);
+                // No FIRST_SWITCHED / flowStartMilliseconds exported: fall back to receipt time (as
+                // sFlow does), honouring the non-null Flow contract and the non-nullable column.
+                return firstSwitched != null ? firstSwitched : receivedAt;
             }
 
             @Override
@@ -91,7 +94,7 @@ public class Netflow9FlowBuilder {
                 final var activeTimeout = Optionals.first(raw.FLOW_ACTIVE_TIMEOUT, flowActiveTimeoutFallback).orElse(null);
                 final var inactiveTimeout = Optionals.first(raw.FLOW_INACTIVE_TIMEOUT, flowInactiveTimeoutFallback).orElse(null);
 
-                return new Timeout()
+                final var delta = new Timeout()
                         .withActiveTimeout(activeTimeout)
                         .withInactiveTimeout(inactiveTimeout)
                         .withFirstSwitched(this.getFirstSwitched())
@@ -99,13 +102,18 @@ public class Netflow9FlowBuilder {
                         .withNumBytes(this.getBytes())
                         .withNumPackets(this.getPackets())
                         .calculateDeltaSwitched();
+                // The timeout calc can yield null (no timeouts); default to firstSwitched like the
+                // Flow interface, which is now non-null.
+                return delta != null ? delta : this.getFirstSwitched();
             }
 
             @Override
             public Instant getLastSwitched() {
-                return Optionals.of(raw.LAST_SWITCHED)
+                final var lastSwitched = Optionals.of(raw.LAST_SWITCHED)
                         .map(this.getBootTime()::plus)
                         .orElse(raw.flowEndMilliseconds);
+                // No LAST_SWITCHED / flowEndMilliseconds exported: fall back to receipt time (see getFirstSwitched).
+                return lastSwitched != null ? lastSwitched : receivedAt;
             }
 
             @Override

--- a/src/main/java/org/riptide/flows/parser/netflow9/Netflow9FlowBuilder.java
+++ b/src/main/java/org/riptide/flows/parser/netflow9/Netflow9FlowBuilder.java
@@ -84,9 +84,10 @@ public class Netflow9FlowBuilder {
                 final var firstSwitched = Optionals.of(raw.FIRST_SWITCHED)
                         .map(this.getBootTime()::plus)
                         .orElse(raw.flowStartMilliseconds);
-                // No FIRST_SWITCHED / flowStartMilliseconds exported: fall back to receipt time (as
-                // sFlow does), honouring the non-null Flow contract and the non-nullable column.
-                return firstSwitched != null ? firstSwitched : receivedAt;
+                // No FIRST_SWITCHED / flowStartMilliseconds exported: fall back to the export time (the
+                // packet header timestamp), as goflow2 does — honouring the non-null Flow contract and
+                // the non-nullable column.
+                return firstSwitched != null ? firstSwitched : this.getTimestamp();
             }
 
             @Override
@@ -112,8 +113,8 @@ public class Netflow9FlowBuilder {
                 final var lastSwitched = Optionals.of(raw.LAST_SWITCHED)
                         .map(this.getBootTime()::plus)
                         .orElse(raw.flowEndMilliseconds);
-                // No LAST_SWITCHED / flowEndMilliseconds exported: fall back to receipt time (see getFirstSwitched).
-                return lastSwitched != null ? lastSwitched : receivedAt;
+                // No LAST_SWITCHED / flowEndMilliseconds exported: fall back to the export time (see getFirstSwitched).
+                return lastSwitched != null ? lastSwitched : this.getTimestamp();
             }
 
             @Override

--- a/src/test/java/org/riptide/flows/ipfix/FlowTimeoutTest.java
+++ b/src/test/java/org/riptide/flows/ipfix/FlowTimeoutTest.java
@@ -75,18 +75,21 @@ public class FlowTimeoutTest {
     }
 
     @Test
-    void fallsBackToReceivedAtWhenNoTimingExported() {
+    void fallsBackToExportTimeWhenNoTimingExported() {
         // An exporter that sends no flow-start/end elements: previously getFirstSwitched/
         // getLastSwitched returned null and the flow failed the non-nullable ClickHouse insert.
+        // We fall back to the export (packet header) time, distinct from the receipt time here,
+        // matching goflow2.
+        final var exportTime = Instant.ofEpochSecond(1000);
         final var received = Instant.ofEpochSecond(5000);
         final var raw = new IpfixRawFlow();
-        raw.exportTime = Instant.EPOCH;
+        raw.exportTime = exportTime;
 
         final var flow = new IpFixFlowBuilder(conversionService).buildFlow(received, raw);
 
-        Assertions.assertThat(flow.getFirstSwitched()).isEqualTo(received);
-        Assertions.assertThat(flow.getLastSwitched()).isEqualTo(received);
-        Assertions.assertThat(flow.getDeltaSwitched()).isEqualTo(received);
+        Assertions.assertThat(flow.getFirstSwitched()).isEqualTo(exportTime);
+        Assertions.assertThat(flow.getLastSwitched()).isEqualTo(exportTime);
+        Assertions.assertThat(flow.getDeltaSwitched()).isEqualTo(exportTime);
     }
 
     @Test

--- a/src/test/java/org/riptide/flows/ipfix/FlowTimeoutTest.java
+++ b/src/test/java/org/riptide/flows/ipfix/FlowTimeoutTest.java
@@ -75,6 +75,21 @@ public class FlowTimeoutTest {
     }
 
     @Test
+    void fallsBackToReceivedAtWhenNoTimingExported() {
+        // An exporter that sends no flow-start/end elements: previously getFirstSwitched/
+        // getLastSwitched returned null and the flow failed the non-nullable ClickHouse insert.
+        final var received = Instant.ofEpochSecond(5000);
+        final var raw = new IpfixRawFlow();
+        raw.exportTime = Instant.EPOCH;
+
+        final var flow = new IpFixFlowBuilder(conversionService).buildFlow(received, raw);
+
+        Assertions.assertThat(flow.getFirstSwitched()).isEqualTo(received);
+        Assertions.assertThat(flow.getLastSwitched()).isEqualTo(received);
+        Assertions.assertThat(flow.getDeltaSwitched()).isEqualTo(received);
+    }
+
+    @Test
     void testFirstLastSwitchedValues1() {
         final var raw = new IpfixRawFlow();
         raw.exportTime = Instant.EPOCH;

--- a/src/test/java/org/riptide/flows/netflow9/FlowTimeoutTest.java
+++ b/src/test/java/org/riptide/flows/netflow9/FlowTimeoutTest.java
@@ -41,18 +41,20 @@ public class FlowTimeoutTest {
     }
 
     @Test
-    void fallsBackToReceivedAtWhenNoTimingExported() {
+    void fallsBackToExportTimeWhenNoTimingExported() {
         // A template without FIRST_SWITCHED/LAST_SWITCHED (or flowStart/EndMilliseconds): previously
         // getFirstSwitched/getLastSwitched returned null and the flow failed the non-nullable insert.
+        // We fall back to the header (export) time, distinct from the receipt time here, matching goflow2.
+        final var headerTime = Instant.ofEpochSecond(1000);
         final var received = Instant.ofEpochSecond(5000);
         final var raw = new Netflow9RawFlow();
-        raw.unixSecs = Instant.EPOCH;
+        raw.unixSecs = headerTime;
         raw.sysUpTime = Duration.ZERO;
 
         final var flowMessage = new Netflow9FlowBuilder(valueConversionService).buildFlow(received, raw);
-        Assertions.assertThat(flowMessage.getFirstSwitched()).isEqualTo(received);
-        Assertions.assertThat(flowMessage.getLastSwitched()).isEqualTo(received);
-        Assertions.assertThat(flowMessage.getDeltaSwitched()).isEqualTo(received);
+        Assertions.assertThat(flowMessage.getFirstSwitched()).isEqualTo(headerTime);
+        Assertions.assertThat(flowMessage.getLastSwitched()).isEqualTo(headerTime);
+        Assertions.assertThat(flowMessage.getDeltaSwitched()).isEqualTo(headerTime);
     }
 
     @Test

--- a/src/test/java/org/riptide/flows/netflow9/FlowTimeoutTest.java
+++ b/src/test/java/org/riptide/flows/netflow9/FlowTimeoutTest.java
@@ -41,6 +41,21 @@ public class FlowTimeoutTest {
     }
 
     @Test
+    void fallsBackToReceivedAtWhenNoTimingExported() {
+        // A template without FIRST_SWITCHED/LAST_SWITCHED (or flowStart/EndMilliseconds): previously
+        // getFirstSwitched/getLastSwitched returned null and the flow failed the non-nullable insert.
+        final var received = Instant.ofEpochSecond(5000);
+        final var raw = new Netflow9RawFlow();
+        raw.unixSecs = Instant.EPOCH;
+        raw.sysUpTime = Duration.ZERO;
+
+        final var flowMessage = new Netflow9FlowBuilder(valueConversionService).buildFlow(received, raw);
+        Assertions.assertThat(flowMessage.getFirstSwitched()).isEqualTo(received);
+        Assertions.assertThat(flowMessage.getLastSwitched()).isEqualTo(received);
+        Assertions.assertThat(flowMessage.getDeltaSwitched()).isEqualTo(received);
+    }
+
+    @Test
     void verifyWithActiveTimeout() {
         final var raw = new Netflow9RawFlow();
         raw.unixSecs = Instant.EPOCH;

